lib/addMethod/globalize/before.js

```diff
@@ -5,36 +5,55 @@
 * object - need to ensure that variables are passed and saved correctly in the
 * tests pre substitution.
 */
-var when = require('when');
-var _    = require('lodash');
+const when = require('when');
+const _    = require('lodash');
+
+const localOnly = require('./localOnly');
+
+const REFERENCE_MODIFICATION_ERROR_MESSAGE = 'Modification by reference is deprecated. `before` must return the modified object.';
+
+function validateResult (referencedParams, originalParamsCopy) {
+    /*
+        If function returns undefined, then assume no modification to
+        original params, and so set as default argument
+    */
+    return (returnedParams) => {
+        if (_.isUndefined(returnedParams)) {
+            if (!_.isEqual(referencedParams, originalParamsCopy)) {
+                if (process.env.NODE_ENV === 'development') {
+                    throw new Error(REFERENCE_MODIFICATION_ERROR_MESSAGE);
+                } else {
+                    console.warn(REFERENCE_MODIFICATION_ERROR_MESSAGE);
+                }
+            }
+        } else if (!_.isPlainObject(returnedParams)) {
+            throw new Error('`before` must return an object');
+        }
+        return returnedParams || originalParamsCopy;
+    };
+}
 
-var localOnly = require('./localOnly');
+module.exports = function (config, params = {}) {
 
-module.exports = function (config, params) {
-  var threadneedle = this;
-  return when.promise(function (resolve, reject) {
+    const { _globalOptions } = this;
 
-    when()
+    const originalParams = _.cloneDeep(params);
 
-    // Run global promise first
-    .then(function () {
-        if (_.isFunction(threadneedle._globalOptions.before) && !localOnly(config, 'before')) {
-            return when(threadneedle._globalOptions.before(params));
-        }
-    })
+    //Start by executing globalBefore if provided and globals true
+    const globalBeforeExec = (
+    	_.isFunction(_globalOptions.before) && !localOnly(config, 'before') ?
+    	_globalOptions.before(params) :
```

Comment:
Is there a linter for threadneedle code? Because in this PR there is a mix of tabs and spaces, even on the same line like here. In other areas of the PR this makes the code hard to read as seen in this screenshot.
![image](https://user-images.githubusercontent.com/14856227/64570773-8517f280-d359-11e9-871d-a2278b2d27bb.png)

Reply:
No, there was no linter originally, except the jshint in grunt, but evidently that's not great.

I've already added a eslint for the next commit.